test/common/buffer/buffer_fuzz_test.cc

```diff
@@ -0,0 +1,253 @@
+    for (uint32_t i = 0; i < allocated_slices; ++i) {
+      ::memset(slices[i].mem_, 'c', slices[i].len_);
+      allocated_length += slices[i].len_;
+    }
+    ASSERT(reserve_length <= allocated_length);
+    const uint32_t target_length =
+        std::min(reserve_length, action.reserve_commit().commit_length());
+    uint32_t shrink_length = allocated_length;
+    int32_t shrink_slice = allocated_slices - 1;
+    while (shrink_length > target_length) {
+      ASSERT(shrink_slice >= 0);
+      const uint32_t available = slices[shrink_slice].len_;
+      const uint32_t remainder = shrink_length - target_length;
+      if (available >= remainder) {
+        slices[shrink_slice].len_ -= remainder;
+        break;
+      }
+      shrink_length -= available;
+      slices[shrink_slice--].len_ = 0;
+    }
+    target_buffer.commit(slices, allocated_slices);
+    ASSERT(previous_length + target_length == target_buffer.length());
+    break;
+  }
+  case test::common::buffer::Action::kCopyOut: {
+    const uint32_t start =
+        std::min(action.copy_out().start(), static_cast<uint32_t>(target_buffer.length()));
+    uint8_t copy_buffer[2 * 1024 * 1024];
+    const uint32_t length =
+        std::min(static_cast<uint32_t>(target_buffer.length() - start),
+                 std::min(action.copy_out().length(), static_cast<uint32_t>(sizeof(copy_buffer))));
+    target_buffer.copyOut(start, length, copy_buffer);
+    const std::string contents = target_buffer.toString();
+    ASSERT(::memcmp(copy_buffer, contents.data() + start, length) == 0);
+    break;
+  }
+  case test::common::buffer::Action::kDrain: {
+    const uint32_t previous_length = target_buffer.length();
+    const uint32_t drain_length =
+        std::min(static_cast<uint32_t>(target_buffer.length()), action.drain());
+    target_buffer.drain(drain_length);
+    ASSERT(previous_length - drain_length == target_buffer.length());
+    break;
+  }
+  case test::common::buffer::Action::kLinearize: {
+    const uint32_t linearize_size =
+        std::min(static_cast<uint32_t>(target_buffer.length()), action.linearize());
+    target_buffer.linearize(linearize_size);
+    Buffer::RawSlice slices[1];
+    const uint64_t slices_used = target_buffer.getRawSlices(slices, 1);
+    if (linearize_size > 0) {
+      ASSERT(slices_used == 1);
+      ASSERT(slices[0].len_ >= linearize_size);
+    }
+    break;
+  }
+  case test::common::buffer::Action::kMove: {
+    const uint32_t source_index = action.move().source_index() % BufferCount;
+    if (target_index == source_index) {
+      break;
+    }
+    Buffer::OwnedImpl& source_buffer = buffers[source_index];
+    if (action.move().length() == 0) {
+      target_buffer.move(source_buffer);
+    } else {
+      target_buffer.move(source_buffer, std::min(static_cast<uint32_t>(source_buffer.length()),
+                                                 action.move().length()));
+    }
+    break;
+  }
+  case test::common::buffer::Action::kRead: {
```

Comment:
Should we verify the read data is what we expect to have read? Same for write below?